inlang/source-code/sdk/src/createNodeishFsWithWatcher.ts

```diff
@@ -8,67 +8,75 @@ import type { NodeishFilesystem } from "@lix-js/fs"
  */
 export const createNodeishFsWithWatcher = (args: {
 	nodeishFs: NodeishFilesystem
-	updateMessages: () => void
+	onChange: () => void
 }): NodeishFilesystem & {
 	stopWatching: () => void
 } => {
-	const pathList: string[] = []
-	let abortControllers: AbortController[] = []
+	const pathList = new Set<string>()
+	const abortControllers = new Set<AbortController>()
 
 	const stopWatching = () => {
 		for (const ac of abortControllers) {
 			ac.abort()
+			abortControllers.delete(ac) // release reference
 		}
-		// release references
-		abortControllers = []
 	}
 
-	const makeWatcher = (path: string) => {
-		;(async () => {
-			try {
-				const ac = new AbortController()
-				abortControllers.push(ac)
-				const watcher = args.nodeishFs.watch(path, {
-					signal: ac.signal,
-					persistent: false,
-				})
-				if (watcher) {
-					//eslint-disable-next-line @typescript-eslint/no-unused-vars
-					for await (const event of watcher) {
-						args.updateMessages()
-					}
+	const makeWatcher = async (path: string) => {
+		try {
+			const ac = new AbortController()
+			abortControllers.add(ac)
+			const watcher = args.nodeishFs.watch(path, {
+				signal: ac.signal,
+				recursive: true,
+				persistent: false,
+			})
+			if (watcher) {
+				//eslint-disable-next-line @typescript-eslint/no-unused-vars
+				for await (const event of watcher) {
+					// whenever the watcher changes we need to update the messages
+					args.onChange()
 				}
-			} catch (err: any) {
-				if (err.name === "AbortError") return
-				// https://github.com/opral/monorepo/issues/1647
-				// the file does not exist (yet)
-				// this is not testable beacause the fs.watch api differs
-				// from node and lix. lenghty
-				else if (err.code === "ENOENT") return
-				throw err
 			}
-		})()
+		} catch (err: any) {
+			if (err.name === "AbortError") return
+			// https://github.com/opral/monorepo/issues/1647
+			// the file does not exist (yet)
+			// this is not testable beacause the fs.watch api differs
+			// from node and lix. lenghty
+			else if (err.code === "ENOENT") return
+			throw err
+		}
 	}
 
-	const readFileAndExtractPath = (path: string, options: { encoding: "utf-8" | "binary" }) => {
-		if (!pathList.includes(path)) {
-			makeWatcher(path)
-			pathList.push(path)
+	/**
+	 * Creates watchers on-the-fly for any file or directory that is not yet watched.
```

Comment:
@LorisSigrist  please add a comment that the only reason we have to do this is that node 18 on linux does not support recursive file watching, otherwise we would add a recursive watch at the root and then just filter out the events with the list of used files.